xen/arch/arm/coproc/coproc.c

```diff
@@ -73,162 +69,293 @@ static struct coproc_device *coproc_find_by_path(const char *path)
+    xfree(vcoproc);
+    return ERR_PTR(ret);
 }
 
 static void coproc_deinit_vcoproc(struct vcoproc_instance *vcoproc)
 {
-    struct coproc_device *coproc;
+    struct mcoproc_device *mcoproc;
 
     if ( !vcoproc )
         return;
 
-    coproc = vcoproc->coproc;
-    spin_lock(&coproc->vcoprocs_lock);
+    mcoproc = vcoproc->mcoproc;
+    spin_lock(&mcoproc->vcoprocs_lock);
     list_del(&vcoproc->vcoproc_elem);
-    spin_unlock(&coproc->vcoprocs_lock);
-    coproc->ops->vcoproc_deinit(vcoproc);
+    spin_unlock(&mcoproc->vcoprocs_lock);
+    mcoproc->ops->vcoproc_deinit(vcoproc);
     xfree(vcoproc);
 }
 
-static int coproc_attach_to_domain(struct domain *d,
-                                   struct coproc_device *coproc)
+static int vcoproc_spawn(struct domain *d,
+                         struct mcoproc_device *mcoproc,
+                         const struct dt_device_node *np_vcoproc)
 {
     struct vcoproc *vcoproc_d = &d->arch.vcoproc;
     struct vcoproc_instance *vcoproc;
     int ret;
 
-    if ( !coproc )
+    if ( !mcoproc )
         return -EINVAL;
 
     spin_lock(&coprocs_lock);
 
-    if ( coproc_is_created_vcoproc(d, coproc) )
-    {
-        ret = -EEXIST;
-        goto out;
-    }
-
-    vcoproc = coproc_init_vcoproc(d, coproc);
+    vcoproc = coproc_init_vcoproc(d, mcoproc, np_vcoproc);
     if ( IS_ERR(vcoproc) )
     {
+        printk("coproc_init_vcoproc failed\n");
         ret = PTR_ERR(vcoproc);
         goto out;
     }
 
-    ret = vcoproc_scheduler_vcoproc_init(coproc->sched, vcoproc);
+    ret = vcoproc_scheduler_vcoproc_init(mcoproc->sched, vcoproc);
     if ( ret )
     {
         coproc_deinit_vcoproc(vcoproc);
         goto out;
     }
 
-    BUG_ON(vcoproc_d->num_instances >= num_coprocs);
     list_add_tail(&vcoproc->instance_elem, &vcoproc_d->instances);
     vcoproc_d->num_instances++;
 
-    printk("Created vcoproc \"%s\" for dom%u\n",
-           dev_path(coproc->dev), d->domain_id);
+    printk("Spawned vcoproc \"%s\" for dom%u\n",
+           dt_node_full_name(np_vcoproc), d->domain_id);
 
 out:
     spin_unlock(&coprocs_lock);
 
     return ret;
 }
 
-static int coproc_find_and_attach_to_domain(struct domain *d, const char *path)
+static int vcoproc_spawn_path(struct domain *d, const char *path,
+                              const struct dt_device_node *np_vcoproc)
 {
-    struct coproc_device *coproc;
+    struct mcoproc_device *mcoproc;
 
-    coproc = coproc_find_by_path(path);
-    if ( !coproc )
+    mcoproc = mcoproc_find_by_path(path);
+    if ( !mcoproc )
+    {
+        printk("coproc %s is not found\n", path);
         return -ENODEV;
+    }
+
+    return vcoproc_spawn(d, mcoproc, np_vcoproc);
+}
+
+static int vcoproc_make_node(const struct domain *d, void *fdt,
+                             const struct dt_device_node *cnode,
+                             const struct dt_device_node *vnode)
+{
+    const struct dt_property *cprop, *vprop;
+    const char *name, *path;
+    int res = 0;
```

Comment:
That is the function based on a code taken from a different file, as well as next one. Will make the stuff consistent across this file.